modules/effects/arrow-effects/src/main/kotlin/arrow/effects/internal/IOConnection.kt

```diff
@@ -0,0 +1,112 @@
+package arrow.effects.internal
+
+import arrow.effects.ForIO
+import arrow.effects.IO
+import arrow.effects.fix
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Represents a composite of functions (meant for cancellation) that are stacked. cancel() is idempotent,
+ * and all methods are thread-safe & atomic.
+ */
+sealed class IOConnection {
+
+  /**
+   * Cancels the unit of work represented by this reference.
+   *
+   * Guaranteed idempotency - calling it multiple times should have the same side-effect as calling it only
+   * once. Implementations of this method should also be thread-safe.
+   */
+  abstract fun cancel(): CancelToken<ForIO>
+
+  abstract fun isCanceled(): Boolean
+
+  /**
+   * Pushes a cancelable reference on the stack, to be popped or canceled later in FIFO order.
+   */
+  abstract fun push(token: CancelToken<ForIO>): Unit
```

Comment:
I understand the semantics behind this function name. But, for the user's sake, I'd replace it with something more like `push` -> `register` and `pop` -> `unregisterLatest`

Reply:
More inline with `setDisposable` and `setCancellable` here http://reactivex.io/RxJava/2.x/javadoc/io/reactivex/ObservableEmitter.html#setDisposable-io.reactivex.disposables.Disposable-

Reply:
Actually this is not a stack, but a queue. It's **FIFO**, so `unregisterLatest` wouldn't be accurate either. Ideas?